File: advent-of-code/src/day01/day-01.rb
#!/usr/bin/env ruby

class Santa
  def initialize()
    @floor = 0
    @step = 0
    @entered_basement = 0
  end
  def followInstruction(i)
    @step += 1
    if ( i == '(' )
      @floor += 1
    elsif ( i == ')' )
      @floor -= 1
      if ( @entered_basement == 0 and @floor < 0 )
        @entered_basement = @step
      end
    end
  end
  def instructions( s )
    s.split("").each do |c|
      self.followInstruction(c)
    end
  end
  def onFloor
    @floor
  end
  def enteredBasement
    @entered_basement
  end
end

data = File.open( "day-01.txt", "r" ).read

santa = Santa.new
santa.instructions( data )
print "Santa ended up on floor #{santa.onFloor}\n"
print "Santa entered the basement on step #{santa.enteredBasement}\n"

File: advent-of-code/src/day01/day-01.php
#!/usr/bin/env php
<?php

$data = file_get_contents( "day-01.txt" );

$floor = 0;
$entered_basement = 0;

$l = strlen( $data );
for ( $i = 0 ; $i < $l ; $i++ ) {
  if ( $data[$i] == '(' ) {
    $floor += 1;
  } else if ( $data[$i] == ')' ) {
    $floor -= 1;
    if ( $entered_basement == 0 && $floor < 0 ) {
      $entered_basement = $i + 1;
    }
  }
}

echo "Santa ended up on floor: ", $floor, "\n";
echo "Santa first entered the basement on step: ", $entered_basement, "\n";
